src/sql/SqlService.ts

```diff
@@ -0,0 +1,408 @@
+        } catch (error) {
+            throw new IllegalArgumentError(`Invalid argument given to execute(): ${error.message}`, error)
+        }
+
+        const connection = this.connectionRegistry.getRandomConnection(true);
+        if (connection === null) {
+            throw new HazelcastSqlException(
+                this.connectionManager.getClientUuid(),
+                SqlErrorCode.CONNECTION_PROBLEM,
+                'Client is not currently connected to the cluster.'
+            );
+        }
+
+        const queryId = SqlQueryId.fromMemberId(connection.getRemoteUuid());
+
+        const expectedResultType: SqlExpectedResultType =
+            sqlStatement.options?.expectedResultType ? SqlExpectedResultType[sqlStatement.options.expectedResultType]
+                : SqlServiceImpl.DEFAULT_EXPECTED_RESULT_TYPE;
+
+        let timeoutMillis: Long;
+        if(sqlStatement.options?.timeoutMillis){
+            if(typeof sqlStatement.options.timeoutMillis === 'number'){
+                timeoutMillis = Long.fromNumber(sqlStatement.options.timeoutMillis);
+            }else timeoutMillis = sqlStatement.options.timeoutMillis;
+        }else timeoutMillis = SqlServiceImpl.DEFAULT_TIMEOUT;
+
+        try {
+            const serializedParams = [];
+            if (Array.isArray(sqlStatement.params)) { // params can be undefined
+                for (let param of sqlStatement.params) {
+                    param = this.convertToStringIfDatetimeValue(param);
+                    serializedParams.push(this.serializationService.toData(param));
+                }
+            }
+            const cursorBufferSize = sqlStatement.options?.cursorBufferSize ?
+                sqlStatement.options.cursorBufferSize : SqlServiceImpl.DEFAULT_CURSOR_BUFFER_SIZE;
+            const requestMessage = SqlExecuteCodec.encodeRequest(
+                sqlStatement.sql,
+                serializedParams,
+                timeoutMillis,
+                cursorBufferSize,
+                sqlStatement.options?.schema || sqlStatement.options?.schema === '' ?
+                    sqlStatement.options.schema : SqlServiceImpl.DEFAULT_SCHEMA,
+                expectedResultType,
+                queryId
+            );
+
+            const res = SqlResultImpl.newResult(
+                this,
+                this.serializationService,
+                connection,
+                queryId,
+                cursorBufferSize,
+                sqlStatement.options?.hasOwnProperty('returnRawResult') ?
+                    sqlStatement.options.returnRawResult : SqlServiceImpl.DEFAULT_FOR_RETURN_RAW_RESULT,
+                this.connectionManager.getClientUuid()
+            );
+
+            this.invocationService.invokeOnConnection(connection, requestMessage).then(clientMessage => {
+                this.handleExecuteResponse(clientMessage, res);
+            }).catch(err => {
+                if(!connection.isAlive()){
```

Comment:
Can we extract this to a separate function, and use it in other places? I see that Java does such rethrow logic for here, for handleExecute/FetchResponse and for close